File: leetcode/c++/0057-insert-interval.cc
// https://leetcode.com/problems/insert-interval/
// Medium

#include<vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> ret;
        bool inserted = false;

        for (vector<int> &curr: intervals) {
            if (inserted || curr[1] < newInterval[0]) {
                ret.push_back(curr);
            } else if (newInterval[1] < curr[0]) {
                ret.push_back(newInterval);
                ret.push_back(curr);
                inserted = true;
            } else {
                newInterval[0] = min(curr[0], newInterval[0]);
                newInterval[1] = max(curr[1], newInterval[1]);
            }
        }
        if (!inserted) ret.push_back(newInterval);
        return ret;
    }
};

File: leetcode/python/0057-insert-interval.py
# https://leetcode.com/problems/insert-interval/
# Medium

# Key is to change newInterval so that it keeps track of the largest interval so far
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ret = []
        for i in intervals:
            if i[1] < newInterval[0]:
                ret.append(i)
            elif newInterval[1] < i[0]:
                ret.append(newInterval)
                newInterval = i
            else:
                newInterval[0] = min(i[0], newInterval[0])
                newInterval[1] = max(i[1], newInterval[1])

        ret.append(newInterval)
        return ret
        